contracts/sol4/bridgeReserves/uniswapCurve/KyberUniswapCurveReserve.sol

```diff
@@ -0,0 +1,458 @@
+        }
+    }
+
+    function doTradeEthToToken(
+        ERC20 token,
+        ERC20 bridgeToken,
+        bool useCurve,
+        uint256 srcAmount
+    )
+        internal returns(uint destAmount)
+    {
+        address[] memory path = new address[](2);
+        path[0] = weth;
+        if (!useCurve) {
+            // directly swap with Uniswap
+            path[1] = address(token);
+            uniswapRouter.swapExactETHForTokens.value(srcAmount)(
+                0, path, address(this), DEADLINE
+            );
+        } else {
+            // swap eth -> bridge token on Uniswap
+            path[1] = address(bridgeToken);
+            uniswapRouter.swapExactETHForTokens.value(srcAmount)(
+                0, path, address(this), DEADLINE
+            );
+            // swap bridge token -> dest on Curve
+            CurveDefiInterface(curveDefiAddress[bridgeToken]).exchange(
+                tokenIndex[bridgeToken],
+                tokenIndex[token],
+                bridgeToken.balanceOf(address(this)),
+                0
+            );
+        }
+        destAmount = token.balanceOf(address(this));
+    }
+
+    function doTradeTokenToEth(
+        ERC20 token,
+        ERC20 bridgeToken,
+        bool useCurve,
+        uint256 srcAmount
+    )
+        internal returns(uint destAmount)
+    {
+        address[] memory path = new address[](2);
+        path[1] = weth;
+        if (!useCurve) {
+            // directly swap with Uniswap
+            path[0] = address(token);
+            uniswapRouter.swapExactTokensForETH(
+                srcAmount, 0, path, address(this), DEADLINE
+            );
+        } else {
+            // swap from src -> bridge token on Curve
+            CurveDefiInterface(curveDefiAddress[bridgeToken]).exchange(
+                tokenIndex[token],
+                tokenIndex[bridgeToken],
+                srcAmount,
+                0
+            );
+            // can't trust Curve's returned destQty
+            uint256 destQty = bridgeToken.balanceOf(address(this));
+            // swap from bridge token -> eth on Uniswap
+            path[0] = address(bridgeToken);
+            uniswapRouter.swapExactTokensForETH(
+                destQty, 0, path, address(this), DEADLINE
+            );
+        }
+        destAmount = address(this).balance;
+    }
+
+    function getUniswapDestAmount(
+        ERC20 token,
+        uint256 srcQty,
+        bool ethToToken
+    ) internal view returns (uint256 destAmount) {
+        address pair = uniswapPair[token];
+        if (pair == address(0)) { return 0; }
+        uint256 wethBalance = ERC20(weth).balanceOf(pair);
+        uint256 tokenBalance = token.balanceOf(pair);
+        if (ethToToken) {
+            destAmount = uniswapRouter.getAmountOut(srcQty, wethBalance, tokenBalance);
+        } else {
+            destAmount = uniswapRouter.getAmountOut(srcQty, tokenBalance, wethBalance);
+        }
+    }
+
+    function getCurveDestAmount(
```

Comment:
above it says the returned dest amount is not accurate.

is this value accurate?

Reply:
yes, it is get func, they does return the correct value (i.e destAmount), for exchange func, it doesn't